File: JavaScript/Data Structure/Hash Map/Array Implementation/arrayHashMap.js
class HashMap {
    constructor(size=10) {
        this.size = size;
        this.array = Array.from({ length: size }, () => []);
    }

    static hash(key) {
        let hash = 0;
        for (let i = 0; i < key.length; i++) {
            hash = ((hash << 5) - hash) + key.charCodeAt(i);
            hash = hash & 0xffffffff;
        }
        return Math.abs(hash);
    }

    set(key, value) {
        const hashedKey = HashMap.hash(key);
        const index = hashedKey % this.size;

        for(let i = 0; i < this.array[index].length; i++) {
            if(this.array[index][i] === key) {
                this.array[index][i] = [key, value];
                return this;
            }
        }
        this.array[index].push([key, value]);
        return this;
    }

    get(key) {
        const hashedKey = HashMap.hash(key);
        let index = hashedKey % this.size;

        for(let i = 0; i < this.array[index].length; i++) {
            if(this.array[index][i][0] == key) {
                return this.array[index][i][1];
            }
        }
        return null;
    }
}

const hashMap = new HashMap(10);

const data = [
    ['Name', 'Luffy'],
    ['Occupation', 'Pirate'],
    ['Age', 20],
    ['Favourite Food', 'Meat'],
    ['Devil Fruit User', true], 
    ['Devil Fruit Name', "Gomu Gomu No Mi"]
];

for (let i = 0; i < data.length; i++) {
    const [key, value] = data[i];
    hashMap.set(key, value);
}

for (let i = 0; i < data.length; i++) {
    const [key, value] = data[i];
    console.log(hashMap.get(key));
}

console.log(hashMap.array);

File: Python/Data Structure/Hash Map/Array-Implementation/array_hash_map.py
class HashMap:
    def __init__(self, size=10):
        self.size = size
        self.array = [[] for _ in range(size)]
    
    def set(self, key, value):
        hashed_key = hash(key)
        index = hashed_key % self.size

        for i, (k, v) in enumerate(self.array[index]):
            if k == key:
                self.array[index][i] = (key, value)
                return

        self.array[index].append((key, value))
        return
    
    def get(self, key):
        hashed_key = hash(key)
        index = hashed_key % self.size

        for k, v in self.array[index]:
            if k == key:
                return v

        return None

hash_map = HashMap()


items = [("Name", "Luffy"), ("Occupation", "Pirate"), ("Age", 20), ("Favourite Food", "Meat"), ("Devil Fruit User", True)]

for key, value in items:
    hash_map.set(key, value)

print(hash_map.array)

print(hash_map.get("Devil Fruit User"))